libsql-wal/src/storage/backend/s3.rs

```diff
@@ -153,14 +166,19 @@ impl<IO: Io> S3Backend<IO> {
         segment_key: &SegmentKey,
     ) -> Result<fst::Map<Vec<u8>>> {
         let s3_index_key = s3_segment_index_key(folder_key, segment_key);
-        let stream = self.s3_get(config, s3_index_key).await?;
-        // TODO: parse header, check if too large to fit memory
-        let bytes = stream
-            .collect()
-            .await
-            .map_err(|e| Error::unhandled(e, ""))?
-            .to_vec();
-        let index = fst::Map::new(bytes).map_err(|_| Error::InvalidIndex)?;
+        let mut stream = self.s3_get(config, s3_index_key).await?.into_async_read();
+        let mut header: SegmentIndexHeader = SegmentIndexHeader::new_zeroed();
+        stream.read_exact(header.as_bytes_mut()).await?;
+        if header.magic.get() != LIBSQL_MAGIC && header.version.get() != 1 {
+            return Err(Error::InvalidIndex("index header magic or version invalid"));
```

Comment:
maybe down the line I feel like we should make these two separate errors so we know which one triggered this.

Reply:
i agree when we have more version we can do that, but right it only means that the segment is corrupted